File: silvestre-wasm/www/src/lib/filters.ts
export type ParamType = "int" | "float" | "select";

export interface ParamSchema {
  key: string;
  label: string;
  type: ParamType;
  min?: number;
  max?: number;
  step?: number;
  default: number | string;
  options?: { label: string; value: string }[];
}

export interface FilterDef {
  name: string;
  label: string;
  category: "effects" | "filters" | "transforms";
  params: ParamSchema[];
}

export const FILTERS: FilterDef[] = [
  // Effects — no params
  { name: "grayscale", label: "Grayscale", category: "effects", params: [] },
  { name: "invert", label: "Invert", category: "effects", params: [] },
  { name: "sepia", label: "Sepia", category: "effects", params: [] },

  // Effects — with params
  {
    name: "brightness",
    label: "Brightness",
    category: "effects",
    params: [
      { key: "delta", label: "Delta", type: "int", min: -255, max: 255, step: 1, default: 30 },
    ],
  },
  {
    name: "contrast",
    label: "Contrast",
    category: "effects",
    params: [
      { key: "factor", label: "Factor", type: "float", min: 0.1, max: 3.0, step: 0.1, default: 1.5 },
    ],
  },

  // Filters — no params
  { name: "sharpen", label: "Sharpen", category: "filters", params: [] },
  { name: "box_blur", label: "Box Blur", category: "filters", params: [] },
  { name: "sobel", label: "Sobel Edge", category: "filters", params: [] },

  // Filters — with params
  {
    name: "gaussian",
    label: "Gaussian Blur",
    category: "filters",
    params: [
      { key: "sigma", label: "Sigma", type: "float", min: 0.1, max: 10.0, step: 0.1, default: 2.0 },
    ],
  },
  {
    name: "median",
    label: "Median",
    category: "filters",
    params: [
      { key: "size", label: "Size", type: "int", min: 3, max: 15, step: 2, default: 3 },
    ],
  },
  {
    name: "canny",
    label: "Canny Edge",
    category: "filters",
    params: [
      { key: "low", label: "Low threshold", type: "float", min: 0.0, max: 100.0, step: 1.0, default: 50.0 },
      { key: "high", label: "High threshold", type: "float", min: 0.0, max: 200.0, step: 1.0, default: 100.0 },
      { key: "sigma", label: "Sigma", type: "float", min: 0.1, max: 5.0, step: 0.1, default: 1.4 },
    ],
  },

  // Transforms
  {
    name: "resize",
    label: "Resize",
    category: "transforms",
    params: [
      { key: "w", label: "Width", type: "int", min: 1, max: 4096, step: 1, default: 512 },
      { key: "h", label: "Height", type: "int", min: 1, max: 4096, step: 1, default: 512 },
    ],
  },
  {
    name: "rotate",
    label: "Rotate",
    category: "transforms",
    params: [
      { key: "angle", label: "Angle (deg)", type: "float", min: -360, max: 360, step: 1, default: 90 },
    ],
  },
  {
    name: "mirror",
    label: "Mirror",
    category: "transforms",
    params: [
      {
        key: "mode",
        label: "Mode",
        type: "select",
        default: "horizontal",
        options: [
          { label: "Horizontal", value: "horizontal" },
          { label: "Vertical", value: "vertical" },
          { label: "Both", value: "both" },
        ],
      },
    ],
  },
  {
    name: "crop",
    label: "Crop",
    category: "transforms",
    params: [
      { key: "x", label: "X", type: "int", min: 0, max: 4096, step: 1, default: 0 },
      { key: "y", label: "Y", type: "int", min: 0, max: 4096, step: 1, default: 0 },
      { key: "w", label: "Width", type: "int", min: 1, max: 4096, step: 1, default: 256 },
      { key: "h", label: "Height", type: "int", min: 1, max: 4096, step: 1, default: 256 },
    ],
  },
];

export const CATEGORIES = ["effects", "filters", "transforms"] as const;

File: silvestre_flutter/example/lib/src/filters.dart
import 'package:flutter/material.dart';

enum FilterCategory { effects, filters, transforms }

enum ParamType { slider, select }

class FilterParam {
  const FilterParam({
    required this.key,
    required this.label,
    required this.type,
    this.min,
    this.max,
    this.step,
    this.defaultValue,
    this.options,
  });

  final String key;
  final String label;
  final ParamType type;
  final double? min;
  final double? max;
  final double? step;
  final double? defaultValue;
  final List<String>? options;
}

class FilterDef {
  const FilterDef({
    required this.name,
    required this.label,
    required this.icon,
    required this.category,
    this.params = const [],
  });

  final String name;
  final String label;
  final IconData icon;
  final FilterCategory category;
  final List<FilterParam> params;
}

const filters = <FilterDef>[
  // Effects
  FilterDef(
    name: 'grayscale',
    label: 'Grayscale',
    icon: Icons.filter_b_and_w,
    category: FilterCategory.effects,
  ),
  FilterDef(
    name: 'invert',
    label: 'Invert',
    icon: Icons.invert_colors,
    category: FilterCategory.effects,
  ),
  FilterDef(
    name: 'sepia',
    label: 'Sepia',
    icon: Icons.filter_vintage,
    category: FilterCategory.effects,
  ),
  FilterDef(
    name: 'brightness',
    label: 'Brightness',
    icon: Icons.brightness_6,
    category: FilterCategory.effects,
    params: [
      FilterParam(
        key: 'delta',
        label: 'Delta',
        type: ParamType.slider,
        min: -255,
        max: 255,
        step: 1,
        defaultValue: 30,
      ),
    ],
  ),
  FilterDef(
    name: 'contrast',
    label: 'Contrast',
    icon: Icons.contrast,
    category: FilterCategory.effects,
    params: [
      FilterParam(
        key: 'factor',
        label: 'Factor',
        type: ParamType.slider,
        min: 0.1,
        max: 3,
        step: 0.1,
        defaultValue: 1.5,
      ),
    ],
  ),

  // Filters
  FilterDef(
    name: 'sharpen',
    label: 'Sharpen',
    icon: Icons.deblur,
    category: FilterCategory.filters,
  ),
  FilterDef(
    name: 'box_blur',
    label: 'Box Blur',
    icon: Icons.blur_on,
    category: FilterCategory.filters,
  ),
  FilterDef(
    name: 'sobel',
    label: 'Sobel Edge',
    icon: Icons.border_style,
    category: FilterCategory.filters,
  ),
  FilterDef(
    name: 'gaussian',
    label: 'Gaussian',
    icon: Icons.blur_circular,
    category: FilterCategory.filters,
    params: [
      FilterParam(
        key: 'sigma',
        label: 'Sigma',
        type: ParamType.slider,
        min: 0.1,
        max: 10,
        step: 0.1,
        defaultValue: 2,
      ),
    ],
  ),
  FilterDef(
    name: 'median',
    label: 'Median',
    icon: Icons.grain,
    category: FilterCategory.filters,
    params: [
      FilterParam(
        key: 'size',
        label: 'Size',
        type: ParamType.slider,
        min: 3,
        max: 15,
        step: 2,
        defaultValue: 3,
      ),
    ],
  ),
  FilterDef(
    name: 'canny',
    label: 'Canny Edge',
    icon: Icons.auto_graph,
    category: FilterCategory.filters,
    params: [
      FilterParam(
        key: 'low',
        label: 'Low',
        type: ParamType.slider,
        min: 0,
        max: 100,
        step: 1,
        defaultValue: 50,
      ),
      FilterParam(
        key: 'high',
        label: 'High',
        type: ParamType.slider,
        min: 0,
        max: 200,
        step: 1,
        defaultValue: 100,
      ),
      FilterParam(
        key: 'sigma',
        label: 'Sigma',
        type: ParamType.slider,
        min: 0.1,
        max: 5,
        step: 0.1,
        defaultValue: 1.4,
      ),
    ],
  ),

  // Transforms
  FilterDef(
    name: 'crop',
    label: 'Crop',
    icon: Icons.crop,
    category: FilterCategory.transforms,
    params: [
      FilterParam(
        key: 'x',
        label: 'X',
        type: ParamType.slider,
        min: 0,
        max: 4096,
        step: 1,
        defaultValue: 0,
      ),
      FilterParam(
        key: 'y',
        label: 'Y',
        type: ParamType.slider,
        min: 0,
        max: 4096,
        step: 1,
        defaultValue: 0,
      ),
      FilterParam(
        key: 'w',
        label: 'Width',
        type: ParamType.slider,
        min: 1,
        max: 4096,
        step: 1,
        defaultValue: 256,
      ),
      FilterParam(
        key: 'h',
        label: 'Height',
        type: ParamType.slider,
        min: 1,
        max: 4096,
        step: 1,
        defaultValue: 256,
      ),
    ],
  ),
  FilterDef(
    name: 'resize',
    label: 'Resize',
    icon: Icons.photo_size_select_large,
    category: FilterCategory.transforms,
    params: [
      FilterParam(
        key: 'w',
        label: 'Width',
        type: ParamType.slider,
        min: 1,
        max: 4096,
        step: 1,
        defaultValue: 512,
      ),
      FilterParam(
        key: 'h',
        label: 'Height',
        type: ParamType.slider,
        min: 1,
        max: 4096,
        step: 1,
        defaultValue: 512,
      ),
    ],
  ),
  FilterDef(
    name: 'rotate',
    label: 'Rotate',
    icon: Icons.rotate_right,
    category: FilterCategory.transforms,
    params: [
      FilterParam(
        key: 'angle',
        label: 'Angle',
        type: ParamType.slider,
        min: -360,
        max: 360,
        step: 1,
        defaultValue: 90,
      ),
    ],
  ),
  FilterDef(
    name: 'mirror',
    label: 'Mirror',
    icon: Icons.flip,
    category: FilterCategory.transforms,
    params: [
      FilterParam(
        key: 'mode',
        label: 'Mode',
        type: ParamType.select,
        options: ['horizontal', 'vertical', 'both'],
      ),
    ],
  ),
];
